Feature: Dashboard Usage Tracking

sinkingpoint opened the pull request:
**What this PR does / why we need it**:
In large grafana deployments, admins are often overwhelmed with
dashboards, harming discoverability and usability in general of Grafana.

In this commit we add in a "visited" column to the dashboard meta, along
side the similar "Created" and "Updated" timestamps. "Visited" gets
updated every time the "GetDashboard" api call is made (But does not
block this call, so as not to break read only deployments). This allows
us to track dashboards that have been abandoned or are no longer in use
and attempt to keep the cruft under control more easily

**Which issue(s) this PR fixes**:
Fixes grafana#16700

**Notes**:

I've added this to the model interface - not sure whether this is the right place for this - perhaps it might be worthwhile adding a new "VisitDashboardCommand" to the Bus or similar? This was the path of least resistance, but happy to change this up if necessary

sinkingpoint commented:
Hummm. Build failure seems... unrelated? ```  1) Inspect drawer tests
 
       Tests various Inspect Drawer scenarios:
 
     Error: The following error originated from your application code, not from Cypress.
 

 
  > Cannot read property 'getText' of null
 

 
TypeError: Cannot read property 'getText' of null
 
    at Object.ai [as getFoldingRanges] (http://localhost:3001/public/build/monaco-json.worker.js:2:215257)
 
    at e.getFoldingRanges (http://localhost:3001/public/build/monaco-json.worker.js:2:221188)
 
    at e.fmr (http://localhost:3001/public/build/monaco-json.worker.js:2:116605)
 
    at e._handleMessage (http://localhost:3001/public/build/monaco-json.worker.js:2:7414)
 
    at Object.handleMessage (http://localhost:3001/public/build/monaco-json.worker.js:2:7018)
 
    at e._handleMessage (http://localhost:3001/public/build/monaco-json.worker.js:2:5038)
 
    at e.handleMessage (http://localhost:3001/public/build/monaco-json.worker.js:2:4606)
 
    at e.onmessage (http://localhost:3001/public/build/monaco-json.worker.js:2:7097)
 
    at Tt.self.onmessage (http://localhost:3001/public/build/monaco-json.worker.js:2:117109)
  ```

marefr reviewed (commented):
Hey, thank you for contributing. Grafana Enterprise includes this in [Usage Insights](https://grafana.com/docs/grafana/latest/enterprise/usage-insights/) and are storing last visited timestamp for a dashboard asynchronously and not in the same database table as dashboards are stored. We don't want to have writes in the read path due to possible performance problems. That's why we cannot accept your currently proposed changes.

marefr commented:
Regarding build failure, think you need to sync with latest master

sinkingpoint commented:
Just to be clear, are you against the principle of this change (And people should upgrade to Grafana enterprise)? Or, if I work this into being async as is apparently done there would this be acceptable?

marefr commented:
> Just to be clear, are you against the principle of this change (And people should upgrade to Grafana enterprise)? 

Not really. But not sure that we in the future maybe make some parts of usage insights an OSS feature and if so maybe better to wait a bit. But the main is that your proposal is not async and are using the dashboard table for storing this, that's a lot of writes if you have a lot of users visiting the same dashboard. Your welcome to work on async solution and we would consider accepting it.

marefr commented:
You can convert PR to a draft until it's ready for review

sinkingpoint commented:
Thanks for the feedback @marefr :) Will work on this now

sinkingpoint committed:
Add dashboard usage accounting backend work

This commit adds the backend work to allow for accounting of when a dashboard
is used/visited. It's organised in such a way that it should be trivial to rip out
in the event that the enterprise Usage Analytics ever gets open sourced.
Nominally, this commit does the following:

 - Adds a database migration to add a table `dashboard_visit` which stores records of when a given user accessed a given database
 - Adds a service which exposes functionality to add entries to this database table
 - Adds an API endpoint /api/dashboard/id/:id/visit which adds an entry, using this service, at the current time for the currently logged in user

sinkingpoint committed:
Add dashboard usage accounting frontend work

Using the previously created API endpoint, this commit adds in an async
call to it during the initDashboard function

sinkingpoint commented:
@marefr I've done a bit more work on this - "visiting" a dashboard is now an extra API call done asynchronously with the fetching of the dashboard

torkelo commented:
Thanks for your contribution! It's a bit tricky how to handle this as it conflicts with the usage tracking in Grafana Enterprise (where we also record this information and have a generic usage tracking API). So this would result in duplicate tracking

edevil commented:
If the problem is the duplicate tracking, how about having this behind a feature flag that gets disabled when using the Enterprise version?

sinkingpoint commented:
CI tests pass now. I'm happy to add a feature flag guard if this is possible / desired

sinkingpoint commented:
@torkelo @marefr Any idea where we go from here?

jessabe commented:
Hej @sinkingpoint -- just a little update that this is being discussed internally right now across a few other teams. But things are a bit slow due to summer. Hopefully we'll get back to you soon on this. Apologies!

sinkingpoint commented:
Thanks @jessabe appreciate the update, and the politics that might be involved :)

mckn commented:
Thanks for opening this PR @sinkingpoint! Unfortunately in its current state we won't be able to approve it. It is a bit too far from being a complete feature and it has been stale too long. Feel free to help out on other parts of Grafana 🙏
